src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderValue.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Formats.Cbor;
+
+namespace System.Security.Cryptography.Cose
+{
+    public readonly struct CoseHeaderValue : IEquatable<CoseHeaderValue>
+    {
+        public readonly ReadOnlyMemory<byte> EncodedValue { get; }
+        private CoseHeaderValue(ReadOnlyMemory<byte> encodedValue)
+        {
+            EncodedValue = encodedValue;
+        }
+
+        public static CoseHeaderValue FromEncodedValue(ReadOnlySpan<byte> encodedValue)
+        {
+            // We don't validate here as we need to know in which label the value is going to be used to validate even more semantics.
+            var encodedValueCopy = new ReadOnlyMemory<byte>(encodedValue.ToArray());
+            CoseHeaderValue value = new CoseHeaderValue(encodedValueCopy);
+            return value;
+        }
+
+        public static CoseHeaderValue FromEncodedValue(byte[] encodedValue)
+            => FromEncodedValue(encodedValue.AsSpan());
+
+        public static CoseHeaderValue FromInt32(int value)
+        {
+            var writer = new CborWriter();
+            writer.WriteInt32(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public static CoseHeaderValue FromString(string value)
+        {
+            var writer = new CborWriter();
+            writer.WriteTextString(value);
+
+            return FromEncodedValue(writer.Encode());
+        }
+
+        public static CoseHeaderValue FromBytes(ReadOnlySpan<byte> value)
+        {
+            var writer = new CborWriter();
+            writer.WriteByteString(value);
+
+            return FromEncodedValue(writer.Encode());
```

Comment:
Would it make sense to stack allocate and/or rent?

1. `writer.Encode()` is going to allocate a `byte[]`.
2. `FromEncodedValue(byte[] encodedValue) calls the `Span<byte>` `FromEncodedValue`.
3. The `Span<byte>` `FromEncodedValue` does a `ToArray` (likely to create a defensive copy?)

So we end up allocating twice writing the CBOR value.

Something like:

```C#
var writer = new CborWriter();
writer.WriteByteString(value);

Span<byte> encoded = writer.BytesWritten <= 256 ? stackalloc byte[256] : new byte[writer.BytesWritten];

int written = writer.Encode(encoded);
CoseHeaderValue result = FromEncodedValue(encoded.Slice(0, written));
return result;
```

I don't know enough context to know how useful or important it is to reduce allocations here.

Reply:
I see, yes, we can improve this method. 
Since we need to allocate the byte[] anyway, I guess it would be better to do:
```cs
byte[] encodedValue = new byte[writer.BytesWritten];
writer.Encode(encodedValue);

return FromEncodedValueInternal(encodedValue); // new factory to avoid creating the defensive copy.
```